api/serializers.py

```diff
@@ -152,6 +153,64 @@ class QuizTitleUpdateSerializer(serializers.Serializer):
     title = serializers.CharField(required=True, allow_blank=False)
 
 
+class AddQuizSessionLogSerializer(serializers.ModelSerializer):
+    quiz_session_code = serializers.CharField(required=True, allow_blank=False)
+    quiz_session_student_id = serializers.IntegerField(required=True, min_value=0)
+    question_multiple_choice_id = serializers.UUIDField(required=False)
+    action = serializers.CharField(required=True)
+
+    class Meta:
+        model = QuizSessionLog
+        fields = [
+            "quiz_session_code",
+            "quiz_session_student_id",
+            "question_multiple_choice_id",
+            "action",
+        ]
+
+    def validate_action(self, value):
+        allowed_actions = ["connected", "disconnected", "reconnected"]
+        if value not in allowed_actions:
+            raise serializers.ValidationError("Invalid action")
+        return value
+
+    def validate(self, data):
+        # resolve objects
+        try:
+            data["quiz_session"] = QuizSession.objects.get(code=data["quiz_session_code"])
+        except QuizSession.DoesNotExist:
+            raise serializers.ValidationError({"quiz_session_code": "Quiz session not found."})
+
+        try:
+            data["quiz_session_student"] = QuizSessionStudent.objects.get(
+                id=data["quiz_session_student_id"]
+            )
+        except QuizSessionStudent.DoesNotExist:
+            raise serializers.ValidationError({"quiz_session_student_id": "Student not found."})
+
+        question_multiple_choice_id = data.get("question_multiple_choice_id")
+        if question_multiple_choice_id:
+            try:
+                data["question_multiple_choice"] = QuestionMultipleChoice.objects.get(
+                    id=question_multiple_choice_id
+                )
+            except QuestionMultipleChoice.DoesNotExist:
+                raise serializers.ValidationError(
+                    {"question_multiple_choice_id": "Question not found."}
+                )
+        else:
+            data["question_multiple_choice"] = None
+
+        return data
+
+    def create(self, validated_data):
+        validated_data.pop("quiz_session_code")
```

Comment:
Removing this means that you won't be saving the code; we shouldn't remove this.